### PR TITLE
US Regression Test Optimization

### PR DESCRIPTION
**Pull request by ozmorph**
Following up on [my comment](https://github.com/mrc-ide/covid-sim/pull/366#pullrequestreview-428065148) in #366 , I found low-hanging fruit inside the `AssignPeopleToPlaces()` function and MicroCellPosition code that results in a substantial (~42.8%) run-time reduction for the US regression test.

Merging this PR would allow for more frequent local testing.

Running `time ./regressiontest_US_based.py` on `master` on my workstation:

```
real    4m24.706s
user    4m52.256s
sys     0m7.381s
```

Running the same test on this PR:

```
real    2m31.731s
user    2m58.940s
sys     0m7.318s
```

### CODE REVIEW

**Comment by ozmorph** [comment]
Weird that it fails on Windows. I'll check locally with a MSVC compile tomorrow.

**Comment by ozmorph** [comment]
Local MSVC test was successful, perhaps a fluke with the container. Rebased with master to trigger another test run.

**Comment by ozmorph** [comment]
> Local MSVC test was successful, perhaps a fluke with the container. Rebased with master to trigger another test run.

Found the issue, Windows regression test appears to be proceeding forward. I had only reserved space in the `mcell_country` vector and not initialized it. So in the proceeding lines when it performed an array access to initialize each element, it was reaching into uninitialized memory. Windows is much more particular about out-of-bounds memory access than OSX or Linux, so this explains why Linux/OSX were passing and Windows wasn't.

File: src/MicroCellPosition.hpp [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
-	MicroCellPosition operator+(Direction direction) const;
-	MicroCellPosition& operator+=(Direction direction);
+	inline MicroCellPosition& operator+=(Direction direction)
+	{
+		switch (direction) {
+			case Right: this->x += 1; break;
+			case Up: this->y -= 1; break;
+			case Left: this->x -= 1; break;
+			case Down: this->y += 1; break;
+			default: throw std::out_of_range("direction");
```
Comment (matt-gretton-dann) [inline src/MicroCellPosition.hpp:19]: Exceptions and OpenMP don't play well together.  Convert this into an ERR_CRITICAL please.

**Comment by ozmorph** [comment]
Since opening this PR, I have made several additional optimizations. I reran the same test between `master` and this branch to get updated numbers. 44% run-time reduction.

```bash
ozmorph@home:~/covid-sim/build$ make test ARGS="-j6 V"
Running tests...
Test project /home/ozmorph/covid-sim/build
    Start 1: inttest-us-based-j1
    Start 2: inttest-us-based-j2
1/2 Test #2: inttest-us-based-j2 ..............   Passed  146.94 sec
2/2 Test #1: inttest-us-based-j1 ..............   Passed  170.99 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) = 170.99 sec
```

vs.

```bash
ozmorph@home:~/covid-sim/build$ make test ARGS="-j6 V"
Running tests...
Test project /home/ozmorph/covid-sim/build
    Start 1: inttest-us-based-j1
    Start 2: inttest-us-based-j2
1/2 Test #2: inttest-us-based-j2 ..............   Passed   74.01 sec
2/2 Test #1: inttest-us-based-j1 ..............   Passed   95.28 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) =  95.28 sec
```

**Comment by ozmorph** [comment]
> Mostly OK. What is the change in performance for the four tests (UK, US) x (j1, j2) (see updated tests on master)? Does this also improve on Linux? (I'm not too worried about as macOS).
> 
> Whilst it is significant in the integration test loop the timing for the model setup is overall a small part of the run. Running the models is what happens multiple times - so any improvements there will be more significant than shown in the integration tests.

I have started a run with all tests enabled and will respond shortly.

**Comment by ozmorph** [comment]
> > Mostly OK. What is the change in performance for the four tests (UK, US) x (j1, j2) (see updated tests on master)? Does this also improve on Linux? (I'm not too worried about as macOS).
> > Whilst it is significant in the integration test loop the timing for the model setup is overall a small part of the run. Running the models is what happens multiple times - so any improvements there will be more significant than shown in the integration tests.
> 
> I have started a run with all tests enabled and will respond shortly.

@matt-gretton-dann 

I do not have sufficient memory on my current workstation (16 GB) to run all 4 tests at the same time. I am running the UK tests separately like I did with the US test.

In addition, I do recognize that the changes in this particular PR do not address code outside of model setup. So I will not be surprised to see only minor improvements to run-time for the UK tests. However as stated in #366, I have found locations within `Sweep.cpp` and `Update.cpp` that are ripe for better CPU caching behavior. But the changes to structures like `Person` and `Household` require significant refactors in the rest of the code that I thought would be better done in separate pull requests. I will address this in a new PR/Issue soon.

**Comment by ozmorph** [comment]
Looks like one of the last 3 commits is causing a failure on the UK tests. I'm diagnosing now.

File: src/MicroCellPosition.hpp (hunk 1 above)
Comment (ozmorph) [reply src/MicroCellPosition.hpp]: Done.

**Comment by ozmorph** [comment]
Regression test fixed. Waiting on test run from `master` before posting run-time improvements for UK tests.

**Comment by ozmorph** [comment]
@matt-gretton-dann Running the UK tests, there is a ~14% run-time reduction between `master` and this PR.

```bash
ozmorph@home:~/covid-sim/build$ make test ARGS="-j6 V"
Running tests...
Test project /home/ozmorph/covid-sim/build
    Start 1: inttest-uk-based-j1
    Start 2: inttest-uk-based-j2
1/2 Test #2: inttest-uk-based-j2 ..............   Passed  674.82 sec
2/2 Test #1: inttest-uk-based-j1 ..............   Passed  2805.93 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) = 2805.94 sec
```

vs.

```bash
ozmorph@home:~/covid-sim/build$ make test ARGS="-j6 V"
Running tests...
Test project /home/ozmorph/covid-sim/build
    Start 1: inttest-uk-based-j1
    Start 2: inttest-uk-based-j2
1/2 Test #2: inttest-uk-based-j2 ..............   Passed  547.70 sec
2/2 Test #1: inttest-uk-based-j1 ..............   Passed  2413.15 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) = 2413.16 sec
```

**Comment by ozmorph** [comment]
I have a theory why Windows takes longer than MacOS and Linux. The first run of the single threaded UK integration test on Ubuntu 20.04 uses an average heap allocation of 3.5 GB, I will be doing a thorough heap profile on Windows tomorrow.

Seeing as how the [GitHub Action runners](https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#supported-runners-and-hardware-resources) only have 7 GB RAM total, this would indicate that the integration tests are probably hitting paged memory. I'm not certain yet as to why Linux and MacOS don't see a slow-down, but one theory is that Linux and MacOS containers have a lower RAM requirement for the base OS.

More to follow in perhaps a separate issue. I consider my development on this particular PR done and ready for final review.

**Comment by matt-gretton-dann** [comment]
> I do not have sufficient memory on my current workstation (16 GB) to run all 4 tests at the same time. I am running the UK tests separately like I did with the US test.

Note that the -j option can be dropped to run tests serially: `make test ARGS="-V"` or even `make test`.  This should enable you to run all tests in one go (but serially).  (For reference the ARGS variable takes `ctest` command line options).

> I have a theory why Windows takes longer than MacOS and Linux. The first run of the single threaded UK integration test on Ubuntu 20.04 uses an average heap allocation of 3.5 GB, I will be doing a thorough heap profile on Windows tomorrow.

> Seeing as how the GitHub Action runners only have 7 GB RAM total, this would indicate that the integration tests are probably hitting paged memory. I'm not certain yet as to why Linux and MacOS don't see a slow-down, but one theory is that Linux and MacOS containers have a lower RAM requirement for the base OS.

Possibly - but the Windows tests were still significantly slower when we were running tests serially.  Having just looked at the Actions Virtual Env setup I think Linux runs with a 4G swap space (https://github.com/actions/virtual-environments/pull/965).  I haven't worked this out for Windows/macOS - but I guess for Windows at least it'll be on the temp SSD which will give us somewhere up to 14GB of swap.

**Comment by ozmorph** [comment]
Thanks for the note on the `-j` option!

> Possibly - but the Windows tests were still significantly slower when we were running tests serially. Having just looked at the Actions Virtual Env setup I think Linux runs with a 4G swap space ([actions/virtual-environments#965](https://github.com/actions/virtual-environments/pull/965)). I haven't worked this out for Windows/macOS - but I guess for Windows at least it'll be on the temp SSD which will give us somewhere up to 14GB of swap.

I'm not sure that's necessarily true. The [GitHub action runners](https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#supported-runners-and-hardware-resources) page indicates that 14 GB of SSD disk space is provisioned. However, Microsoft's own [page file documentation](https://docs.microsoft.com/en-us/windows/client-management/determine-appropriate-page-file-size#system-managed-page-files) states that the maximum page file size is 1/8 the size of volume capacity. In this case, the page file would be 2GB which is half of Linux's swap space. There was a recent [Azure Issue](https://github.com/MicrosoftDocs/azure-docs/issues/54770) that pointed out that this could be problematic but had no public resolution.

But as you stated, this does not explain why the regression tests are slower on Windows when ran serially. I'll continue to diagnose.

**Comment by matt-gretton-dann** [comment]
I think I'm OK with this - but I'd like @dlaydon or @weshinsley to comment please.

**Comment by ozmorph** [comment]
@weshinsley Your analysis is correct. I'd be happy to create more in-depth/formal documentation of how I discovered these performance optimizations in the near future, but I'm a bit short for time over the next day or two.

What I can say in short is that I used `perf` on Linux to locate individual lines of code that were causing the biggest performance issues. Then I looked at memory dependencies associated with those lines and determined if small changes to the structs could achieve better CPU cache behavior. The idea behind moving towards parallel structs is not specific to C++, it is quite common in C code as well. https://en.wikipedia.org/wiki/AoS_and_SoA

My intention is to soon create a separate issue that will discuss this analysis technique in more detail and how to best apply it to `Update.cpp` and `Sweep.cpp` in order to further optimize the run-time performance of the code.

Thanks for the review!

**Comment by matt-gretton-dann** [comment]
@ozmorph : Do you mind merging in master/rebasing to resolve conflicts please - then @weshinsley or I will merge.

My working hypothesis is that the PRs that will improve performance most are those like these that changes structure layout/organisation to improve locality of sequential data accesses and have a regular pattern.  This gives the CPU plenty of opportunity to predict correctly and not get blocked on the memory system

**Comment by ozmorph** [comment]
> @ozmorph : Do you mind merging in master/rebasing to resolve conflicts please - then @weshinsley or I will merge.

@weshinsley @matt-gretton-dann: Rebased with master as requested.

> My working hypothesis is that the PRs that will improve performance most are those like these that changes structure layout/organisation to improve locality of sequential data accesses and have a regular pattern. This gives the CPU plenty of opportunity to predict correctly and not get blocked on the memory system

I agree completely!

**Pull request merged**